File: scripts/test_cola_cloud_2.mjs
#!/usr/bin/env node
/**
 * COLA Cloud API test round 2 — match against known wines
 * Uses ~10 of our remaining 495 requests
 */

import { readFileSync, writeFileSync } from 'fs';

const envContent = readFileSync('.env', 'utf8');
const API_KEY = envContent.match(/COLA_CLOUD_API_KEY=(.+)/)?.[1]?.trim();
const BASE = 'https://app.colacloud.us/api/v1';

let requestCount = 0;

async function apiGet(path, params = {}) {
  const url = new URL(`${BASE}${path}`);
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined) url.searchParams.set(k, v);
  }
  console.log(`  GET ${url.pathname}${url.search}`);
  const res = await fetch(url, {
    headers: { 'X-API-Key': API_KEY }
  });
  requestCount++;
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`${res.status}: ${text}`);
  }
  const remaining = res.headers.get('x-ratelimit-remaining');
  if (remaining) console.log(`  → Remaining: ${remaining}`);
  return res.json();
}

async function searchAndDetail(label, searchParams) {
  console.log(`\n=== ${label} ===`);
  const results = await apiGet('/colas', {
    product_type: 'wine',
    per_page: 5,
    approval_date_from: '2015-01-01',
    ...searchParams
  });
  console.log(`  Found: ${results.pagination?.total} COLAs`);

  if (results.data.length === 0) {
    console.log('  (no results)');
    return null;
  }

  // Show search results
  for (const c of results.data.slice(0, 3)) {
    console.log(`  [search] ${c.ttb_id}: ${c.brand_name} — ${c.product_name} | ${c.origin_name} | ABV ${c.abv}%`);
  }

  // Get detail on first result
  const detail = await apiGet(`/colas/${results.data[0].ttb_id}`);
  const d = detail.data;
  console.log(`  [detail] Grapes: ${JSON.stringify(d.grape_varietals)}`);
  console.log(`  [detail] Appellation: ${d.wine_appellation}`);
  console.log(`  [detail] Vintage: ${d.wine_vintage_year}`);
  console.log(`  [detail] Barcode: ${d.barcode_value} (${d.barcode_type})`);
  console.log(`  [detail] Designation: ${d.llm_wine_designation}`);
  console.log(`  [detail] Description: ${d.llm_product_description?.slice(0, 200)}`);
  console.log(`  [detail] Tasting: ${JSON.stringify(d.llm_tasting_note_flavors)}`);
  console.log(`  [detail] Permit: ${d.permit_number} | State: ${d.address_state}`);

  return detail.data;
}

async function main() {
  console.log('\n=== COLA Cloud Test Round 2: Known Wine Matching ===');

  const details = [];

  // Test known producers we have in DB
  // 1. Ridge Monte Bello (US domestic)
  details.push(await searchAndDetail('Ridge Monte Bello', {
    q: 'Ridge Monte Bello'
  }));

  // 2. López de Heredia (Spanish import)
  details.push(await searchAndDetail('López de Heredia', {
    q: 'Lopez de Heredia'
  }));

  // 3. Antinori Tignanello (Italian import)
  details.push(await searchAndDetail('Antinori Tignanello', {
    q: 'Tignanello'
  }));

  // 4. Louis Roederer Cristal (Champagne)
  details.push(await searchAndDetail('Louis Roederer Cristal', {
    q: 'Cristal Roederer'
  }));

  // 5. Château d'Yquem (Sauternes)
  details.push(await searchAndDetail("Château d'Yquem", {
    q: "Yquem"
  }));

  // Test barcode lookup endpoint
  console.log('\n=== Barcode Lookup Test ===');
  // Use the Bessin barcode from test 1
  const barcode = await apiGet('/barcodes/3554770154090');
  console.log(`Barcode lookup result:`);
  console.log(`  Total COLAs with this barcode: ${barcode.data?.total_colas}`);
  console.log(`  COLAs: ${JSON.stringify(barcode.data?.colas?.map(c => `${c.brand_name} ${c.product_name}`))}`);

  // Test: how many recent wine COLAs have grapes on detail?
  // Pull 5 detail records from different origins to check grape coverage
  console.log('\n=== Grape Coverage Spot Check (5 random details) ===');
  const randomSample = await apiGet('/colas', {
    product_type: 'wine',
    per_page: 100,
    approval_date_from: '2024-06-01',
    approval_date_to: '2024-12-31'
  });

  let grapeCount = 0, appellationCount = 0, vintageCount = 0, barcodeCount = 0;
  const detailChecks = randomSample.data.slice(0, 5);
  for (const c of detailChecks) {
    const det = await apiGet(`/colas/${c.ttb_id}`);
    const d = det.data;
    if (d.grape_varietals?.length > 0) grapeCount++;
    if (d.wine_appellation) appellationCount++;
    if (d.wine_vintage_year) vintageCount++;
    if (d.barcode_value) barcodeCount++;
    console.log(`  ${d.brand_name} ${d.product_name}: grapes=${JSON.stringify(d.grape_varietals)} | app=${d.wine_appellation} | vintage=${d.wine_vintage_year} | barcode=${d.barcode_value ? 'yes' : 'no'}`);
  }
  console.log(`\nDetail field coverage (5 samples):`);
  console.log(`  Grapes: ${grapeCount}/5`);
  console.log(`  Appellation: ${appellationCount}/5`);
  console.log(`  Vintage: ${vintageCount}/5`);
  console.log(`  Barcode: ${barcodeCount}/5`);

  // Save all details
  writeFileSync('data/imports/cola_cloud_test2.json', JSON.stringify({
    known_wine_details: details.filter(Boolean),
    barcode_lookup: barcode.data,
    grape_spot_check: detailChecks.map(c => c.ttb_id)
  }, null, 2));

  console.log(`\n✅ Saved to data/imports/cola_cloud_test2.json`);
  console.log(`Total requests this run: ${requestCount}`);
  console.log(`Estimated remaining: ${495 - requestCount}`);
}

main().catch(err => {
  console.error('Error:', err.message);
  process.exit(1);
});

File: pipeline/analyze/test_cola_cloud_2.py
#!/usr/bin/env python3
"""
COLA Cloud API test round 2 -- match against known wines.
Uses ~10 of remaining requests.

Usage:
    python -m pipeline.analyze.test_cola_cloud_2
"""

import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

import httpx

from pipeline.lib.db import get_env

API_KEY = get_env("COLA_CLOUD_API_KEY")
BASE = "https://app.colacloud.us/api/v1"
DATA_DIR = Path(__file__).resolve().parents[2] / "data" / "imports"

request_count = 0


def api_get(path: str, params: dict | None = None) -> dict:
    global request_count
    url = f"{BASE}{path}"
    print(f"  GET {path}")
    r = httpx.get(url, params=params or {}, headers={"X-API-Key": API_KEY}, timeout=30)
    request_count += 1
    r.raise_for_status()
    remaining = r.headers.get("x-ratelimit-remaining")
    if remaining:
        print(f"  -> Remaining: {remaining}")
    return r.json()


def search_and_detail(label: str, search_params: dict) -> dict | None:
    print(f"\n=== {label} ===")
    results = api_get("/colas", {
        "product_type": "wine", "per_page": 5,
        "approval_date_from": "2015-01-01", **search_params,
    })
    print(f"  Found: {results.get('pagination', {}).get('total')} COLAs")

    data = results.get("data", [])
    if not data:
        print("  (no results)")
        return None

    for c in data[:3]:
        print(f"  [search] {c.get('ttb_id')}: {c.get('brand_name')} -- {c.get('product_name')} | {c.get('origin_name')} | ABV {c.get('abv')}%")

    detail = api_get(f"/colas/{data[0]['ttb_id']}")
    d = detail.get("data", {})
    print(f"  [detail] Grapes: {json.dumps(d.get('grape_varietals'))}")
    print(f"  [detail] Appellation: {d.get('wine_appellation')}")
    print(f"  [detail] Vintage: {d.get('wine_vintage_year')}")
    print(f"  [detail] Barcode: {d.get('barcode_value')} ({d.get('barcode_type')})")
    print(f"  [detail] Designation: {d.get('llm_wine_designation')}")
    desc = d.get("llm_product_description") or ""
    print(f"  [detail] Description: {desc[:200]}")

    return d


def main():
    print("\n=== COLA Cloud Test Round 2: Known Wine Matching ===")

    details = []
    details.append(search_and_detail("Ridge Monte Bello", {"q": "Ridge Monte Bello"}))
    details.append(search_and_detail("Lopez de Heredia", {"q": "Lopez de Heredia"}))
    details.append(search_and_detail("Antinori Tignanello", {"q": "Tignanello"}))
    details.append(search_and_detail("Louis Roederer Cristal", {"q": "Cristal Roederer"}))
    details.append(search_and_detail("Chateau d'Yquem", {"q": "Yquem"}))

    # Barcode lookup
    print("\n=== Barcode Lookup Test ===")
    barcode = api_get("/barcodes/3554770154090")
    bc_data = barcode.get("data", {})
    print(f"Barcode lookup result:")
    print(f"  Total COLAs with this barcode: {bc_data.get('total_colas')}")

    # Grape coverage spot check
    print("\n=== Grape Coverage Spot Check (5 random details) ===")
    random_sample = api_get("/colas", {
        "product_type": "wine", "per_page": 100,
        "approval_date_from": "2024-06-01", "approval_date_to": "2024-12-31",
    })

    grape_count = app_count = vintage_count = barcode_count = 0
    for c in random_sample.get("data", [])[:5]:
        det = api_get(f"/colas/{c['ttb_id']}")
        d = det.get("data", {})
        if d.get("grape_varietals") and len(d["grape_varietals"]) > 0:
            grape_count += 1
        if d.get("wine_appellation"):
            app_count += 1
        if d.get("wine_vintage_year"):
            vintage_count += 1
        if d.get("barcode_value"):
            barcode_count += 1
        print(f"  {d.get('brand_name')} {d.get('product_name')}: "
              f"grapes={json.dumps(d.get('grape_varietals'))} | "
              f"app={d.get('wine_appellation')} | vintage={d.get('wine_vintage_year')} | "
              f"barcode={'yes' if d.get('barcode_value') else 'no'}")

    print(f"\nDetail field coverage (5 samples):")
    print(f"  Grapes: {grape_count}/5")
    print(f"  Appellation: {app_count}/5")
    print(f"  Vintage: {vintage_count}/5")
    print(f"  Barcode: {barcode_count}/5")

    # Save
    output = {
        "known_wine_details": [d for d in details if d],
        "barcode_lookup": bc_data,
    }
    out_path = DATA_DIR / "cola_cloud_test2.json"
    out_path.write_text(json.dumps(output, indent=2), encoding="utf-8")
    print(f"\nSaved to {out_path}")
    print(f"Total requests this run: {request_count}")
    print(f"Estimated remaining: {495 - request_count}")


if __name__ == "__main__":
    main()
